design/apistyle-no2.md

```diff
@@ -0,0 +1,496 @@
+
+Further, for more complex `XRDs` (Crossplane `CompositeResourceDefinitions`) the fluent API doesn't provide a clean way to reuse schema prop definitions across multiple `XRDs` like the Cluster `XRD` in the `acme-platform-aws` example, which wraps EKS and Services `XRDs` and associated `Compositions`. In the `Compositions` that map the Cluster `XRD` down to the EKS and Services `XRDs` there is a lot of boilerplate mapping stanzas, which could be automatically mapped if re-using and extending prop definitions across `XRDs` -- even if the `spec` paths change, for example:
+
+```ts
+{
+  fromFieldPath: 'spec.parameters.services.operators.prometheus.version',
+  toFieldPath: 'spec.operators.prometheus.version',
+}
+```
+
+See [#30](https://github.com/crossplane-contrib/crossplane-cdk8s/issues/30).
+
+## Proposal
+
+Provide a more traditional OO API style (no. 2) that maps cleanly to all languages supported by `jsii` and supports `XRD` schema prop reuse across `XRDs` and the option to auto-map fields from a parent `XRD` to a composed `XRD`, even if the `spec` fields change (e.g. remove a portion of the path hierarchy like the Services `XRD` does).
+
+### Authorship Experience
+
+The following sketch illustrates the proposed Dev UX:
+
+This cdk8s Chart would synth the content for the following XRDs and backing `Compositions`:
+
+* `CompositeCluster.aws.platform.acme.io`
+* `EKS.aws.platform.acme.io`
+* `Services.aws.platform.acme.io`
+
+```ts
+import { Chart } from 'cdk8s';
+import { Construct } from 'constructs';
+import { PropType, MetaUI } from 'crossplane-cdk8s';
+import * as crossplane from 'crossplane-cdk8s';
+import * as iam from '../imports/identity.aws.crossplane.io';
+import * as eks from '../imports/eks.aws.crossplane.io';
+import * as helm from '../imports/helm.crossplane.io';
+
+const PATH_METADATA_UID = 'metadata.uid';
+const PATH_METADATA_NAME = 'metadata.name';
+const PATH_METADATA_LABELS = 'metadata.labels';
+const PATH_METADATA_ANNOTATIONS = 'metadata.annotations'
+const PATH_EXTERNAL_NAME = 'metadata.annotations[crossplane.io/external-name]';
+const PATH_CONN_SECRET_NAME = 'spec.writeConnectionSecretToRef.name';
+const PATH_CONN_SECRET_NAMESPACE = 'spec.writeConnectionSecretToRef.namespace';
+
+enum NodeSize {
+  SMALL = 'small',
+  MEDIUM = 'medium',
+  LARGE = 'large',
+}
+
+export class ClusterChart extends Chart {
+
+  private id: StringProp = stringProp({
+    path: 'spec.id',
+    description: 'Cluster ID that other objects will use to refer to this cluster',
+    requireAllInPath: true,
+  });
+
+  private networkRef: StringProp = stringProp({
+    path: 'spec.parameters.networkRef.id',
+    description: 'A reference to the Network object that this Cluster should be connected to.',
+    requireAllInPath: true,
+  });
+
+  private nodeCount: IntegerProp = integerProp({
```

Comment:
Why are these private members instead of local variables in the ctor? Is there any value in accessing them later?

Reply:
Yes, they're shared across builder methods and the generated prop id would be used for auto mapping. 

IMO stylistically it also nice that the chart has the top level schema attributes on the class directly.